Sources/MapboxCoreNavigation/EHorizon/RoadObjectMatcher.swift

```diff
@@ -0,0 +1,116 @@
+        native.matchPolyline(forPolyline: polyline, id: identifier)
+    }
+
+    /**
+     Matches a given polygon to the graph.
+     "Matching" here means we try to find all intersections of the polygon with the road graph
+     and track distances to those intersections as distance to the polygon.
+
+     - parameter polygon: Polygon representing the object.
+     - parameter identifier: Unique identifier of the object.
+     */
+    public func matchPolygon(_ polygon: [CLLocation], identifier: RoadObjectIdentifier) {
+        native.matchPolygon(forPolygon: polygon, id: identifier)
+    }
+
+    /**
+     Matches given gantry (i.e. polyline orthogonal to the road) to the graph.
+     "Matching" here means we try to find all intersections of the gantry with the road graph
+     and track distances to those intersections as distance to the gantry.
+
+     - parameter gantry: Gantry representing the object.
+     - parameter identifier: Unique identifier of the object.
+     */
+    public func matchGantry(_ gantry: [CLLocation], identifier: RoadObjectIdentifier) {
+        native.matchGantry(forGantry: gantry, id: identifier)
+    }
+
+    /**
+     Matches given point to road graph.
+
+     - parameter point: Point representing the object.
+     - parameter identifier: Unique identifier of the object.
+     */
+    public func matchPoint(_ point: CLLocationCoordinate2D, identifier: RoadObjectIdentifier) {
+        native.matchPoint(forPoint: point, id: identifier)
+    }
+
+    /**
+     Cancel road object matching.
+
+     - parameter identifier: Identifier for which matching should be canceled.
+     */
+    public func cancel(identifier: RoadObjectIdentifier) {
+        native.cancel(forId: identifier)
+    }
+
+    init(_ native: MapboxNavigationNative.RoadObjectMatcher) {
+        self.native = native
+    }
+
+    deinit {
+        native.setListenerFor(nil)
+    }
+
+    private let native: MapboxNavigationNative.RoadObjectMatcher
+}
+
+extension RoadObjectMatcher: RoadObjectMatcherListener {
+    public func onRoadObjectMatched(forRoadObject roadObject: MBXExpected<AnyObject, AnyObject>) {
+        if roadObject.isValue(), let value = roadObject.value {
+            guard let roadObject = value as? MapboxNavigationNative.RoadObject else {
+                preconditionFailure("Road object value can't be constructed. Unknown value type.")
+            }
+            delegate?.didMatchRoadObject(result: .success(RoadObject(roadObject)))
+        } else if roadObject.isError(), let error = roadObject.error {
+            guard let error = error as? MapboxNavigationNative.RoadObjectMatcherError else {
+                preconditionFailure("Road object matcher error can't be constructed. Unknown error type.")
+            }
+            delegate?.didMatchRoadObject(result: .failure(RoadObjectMatcherError(error)))
+        } else {
+            preconditionFailure("Expected type is neither a value nor an error.")
+        }
```

Comment:
Looks like a patter which likely to be used in other places as well. I would consider creating a converter from `MBXExpected` to `Swift.Result`. One way of doing this is creating an init method on `Swift.Result`.

Reply:
Nice idea. We'll still need to convert from one `Result` to another, but that's a much more readable code. Another catch is that we'll need to make an `Error`-conformance of native type, but that shouldn't be a big hit on usability.

Reply:
We could implement a more trickery approach by providing "conversion" methods to `Result.init(expected)`, but I don't think it worth it. Your solution is good enough, which can be improved upon if needed.